src/cgutils.cpp

```diff
@@ -919,6 +919,56 @@ static void raise_exception_if(Value *cond, Value *exc, jl_codectx_t *ctx)
                            exc, ctx);
 }
 
+static size_t dereferenceable_size(jl_value_t *jt) {
+    size_t size = 0;
+    if (jl_is_array_type(jt)) {
+        // Array has at least this much data
+        size = sizeof(jl_array_t);
+    } else {
+        size = jl_datatype_size(jt);
+    }
+    return size;
+}
+
+static inline void maybe_mark_argument_dereferenceable(Argument *A, jl_value_t *jt) {
+    if (!jl_is_leaf_type(jt)) {
+        return;
+    }
+    size_t size = dereferenceable_size(jt);
+    if (!size) {
+        return;
+    }
+    llvm::AttrBuilder Attrs;
+    Attrs.addDereferenceableAttr(size);
+#if JL_LLVM_VERSION >= 50000
+    A->addAttr(llvm::AttributeList::get(jl_LLVMContext,
+                                        A->getArgNo() + 1, Attrs));
+#else
+    A->addAttr(llvm::AttributeSet::get(jl_LLVMContext,
+                                        A->getArgNo() + 1, Attrs));
+#endif
```

Comment:
I feel like I don't understand the purpose of AttrBuilder. Wouldn't this be simpler to write as:
```julia
A->addAttr(AttributeSet().addDereferenceableAttr(jl_LLVMContext, A->getArgNo() + 1, size));
```
But that's not really canonical either, since Arguments don't store AttributeSets, so that ends up making several extra copies of the AttributeSet object also. And all that in order to do the actual call:
```julia
A->getParent()->addDereferenceableAttr(A->getArgNo() + 1, size);
```